contracts/axelar-auth-verifier/src/test.rs

```diff
@@ -79,6 +251,155 @@ fn test_transfer_operatorship() {
     assert!(latest_signer_set);
 }
 
+#[test]
+fn test_transfer_operatorship_fail_empty_signer_set() {
+    let (env, _, client) = setup_env();
+
+    let user = Address::generate(&env);
+    let previous_signer_retention = 1;
+
+    initialize(
+        &env,
+        &client,
+        user.clone(),
+        previous_signer_retention,
+        randint(1, 10),
+    );
+
+    let empty_signer_set = generate_empty_signer_set(&env);
+
+    let empty_operator_set = empty_signer_set.to_xdr(&env);
+
+    // should throw an error, empty signer set
+    let res = client.try_transfer_operatorship(&empty_operator_set);
+    assert!(res.is_err());
+}
+
+#[test]
+fn test_transfer_operatorship_fail_zero_weight() {
+    let (env, _, client) = setup_env();
+
+    let user = Address::generate(&env);
+    let previous_signer_retention = 1;
+
+    initialize(
+        &env,
+        &client,
+        user.clone(),
+        previous_signer_retention,
+        randint(1, 10),
+    );
+
+    let new_signers = generate_signer_set(&env, randint(1, 10), true);
+
+    let encoded_new_signer_set = new_signers.signer_set.to_xdr(&env);
+
+    // should throw an error, last signer weight is zero
+    let res = client.try_transfer_operatorship(&encoded_new_signer_set);
+    assert!(res.is_err());
+}
+
+#[test]
+fn test_transfer_operatorship_fail_zero_threshold() {
+    let (env, _, client) = setup_env();
+
+    let user = Address::generate(&env);
+    let previous_signer_retention = 1;
+
+    initialize(
+        &env,
+        &client,
+        user.clone(),
+        previous_signer_retention,
+        randint(1, 10),
+    );
+
+    let mut new_signers = generate_signer_set(&env, randint(1, 10), false);
+
+    // set the threshold to zero
+    new_signers.signer_set.threshold = U256::from_u32(&env, 0);
+
+    let encoded_new_signer_set = new_signers.signer_set.to_xdr(&env);
+
+    // should error because the threshold is set to zero
+    let res = client.try_transfer_operatorship(&encoded_new_signer_set);
+    assert!(res.is_err());
+}
+
+#[test]
+fn test_transfer_operatorship_fail_low_total_weight() {
+    let (env, _, client) = setup_env();
+
+    let user = Address::generate(&env);
+    let previous_signer_retention = 1;
+
+    initialize(
+        &env,
+        &client,
+        user.clone(),
+        previous_signer_retention,
+        randint(1, 10),
+    );
+
+    let mut new_signers = generate_signer_set(&env, randint(1, 10), false);
+
+    let env = &env;
+    let zero = U256::from_u32(env, 0);
+    let one = U256::from_u32(env, 1);
+    let mut total_weight = zero.clone();
+
+    for weight in new_signers.signer_set.signers.iter().map(|s| s.1) {
```

Comment:
try to avoid using for loops whenever possible. use the reduce/fold functional pattern to accumulate
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold

the extra `map` also isn't needed. you can destructure the argument in the fold itself

Reply:
changed